### PR TITLE
Fallback support for failed metadata retrieval (Fixes #13).

### LINKED ISSUE #13: handle doi requests for articles with incomplete metadata
If I try something that is new or very old, I hit an error that I guess is needing to handle zero citation data as a special case?
```julia 
using ShortCodes
DOI("10.1103/PhysRev.65.117")

Failed to show value:

BoundsError: attempt to access 0-element JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}} at index [1]

throw_boundserror(::JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}}, ::Tuple{Int64})@abstractarray.jl:703
checkbounds@abstractarray.jl:668[inlined]
getindex(::JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}}, ::Int64)@JSON3.jl:158
var"##fetch_metadata_unmemoized"(::String)@doi.jl:68
#5@Memoize.jl:62[inlined]
get!@iddict.jl:178[inlined]
fetch_metadata@Memoize.jl:61[inlined]
var"##fetch_metadata_unmemoized"(::ShortCodes.DOI{String})@doi.jl:63
#3@Memoize.jl:62[inlined]
get!(::ShortCodes.var"#3#4"{ShortCodes.DOI{String}}, ::IdDict{Any, Any}, ::Any)@iddict.jl:178
fetch_metadata(::ShortCodes.DOI{String})@Memoize.jl:61
getproperty(::ShortCodes.DOI{String}, ::Symbol)@doi.jl:26
emph_author(::ShortCodes.DOI{String})@doi.jl:104
show(::IOContext{IOBuffer}, ::MIME{Symbol("text/html")}, ::ShortCodes.DOI{String})@doi.jl:43
show_richest(::IOContext{IOBuffer}, ::Any)@PlutoRunner.jl:1186
show_richest_withreturned@PlutoRunner.jl:1100[inlined]
format_output_default(::Any, ::Any)@PlutoRunner.jl:1000
#format_output#61@PlutoRunner.jl:1017[inlined]
#53@PlutoRunner.jl:909[inlined]
var"#with_io_to_logs#115"(::Bool, ::Base.CoreLogging.LogLevel, ::typeof(Main.PlutoRunner.with_io_to_logs), ::Main.PlutoRunner.var"#53#56"{Module, ShortCodes.DOI{String}, Dict{Tuple{UInt64, Int64}, Int64}})@PlutoRunner.jl:2424
(::Main.PlutoRunner.var"#119#120"{Bool, Base.CoreLogging.LogLevel, Main.PlutoRunner.var"#53#56"{Module, ShortCodes.DOI{String}, Dict{Tuple{UInt64, Int64}, Int64}}})()@PlutoRunner.jl:2448
with_logstate(::Function, ::Any)@logging.jl:511
with_logger@logging.jl:623[inlined]
#with_logger_and_io_to_logs#118@PlutoRunner.jl:2447[inlined]
var"#formatted_result_of#50"(::Bool, ::typeof(Main.PlutoRunner.formatted_result_of), ::Base.UUID, ::Base.UUID, ::Bool, ::Vector{String}, ::Nothing, ::Module)@PlutoRunner.jl:908
top-level scope@WorkspaceManager.jl:527

DOI("10.1103/PhysRevA.107.010101")

Failed to show value:

BoundsError: attempt to access 0-element JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}} at index [1]

throw_boundserror(::JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}}, ::Tuple{Int64})@abstractarray.jl:703
checkbounds@abstractarray.jl:668[inlined]
getindex(::JSON3.Array{Union{}, Vector{UInt8}, Vector{UInt64}}, ::Int64)@JSON3.jl:158
var"##fetch_metadata_unmemoized"(::String)@doi.jl:68
#5@Memoize.jl:62[inlined]
get!@iddict.jl:178[inlined]
fetch_metadata@Memoize.jl:61[inlined]
var"##fetch_metadata_unmemoized"(::ShortCodes.DOI{String})@doi.jl:63
#3@Memoize.jl:62[inlined]
get!(::ShortCodes.var"#3#4"{ShortCodes.DOI{String}}, ::IdDict{Any, Any}, ::Any)@iddict.jl:178
fetch_metadata(::ShortCodes.DOI{String})@Memoize.jl:61
getproperty(::ShortCodes.DOI{String}, ::Symbol)@doi.jl:26
emph_author(::ShortCodes.DOI{String})@doi.jl:104
show(::IOContext{IOBuffer}, ::MIME{Symbol("text/html")}, ::ShortCodes.DOI{String})@doi.jl:43
show_richest(::IOContext{IOBuffer}, ::Any)@PlutoRunner.jl:1186
show_richest_withreturned@PlutoRunner.jl:1100[inlined]
format_output_default(::Any, ::Any)@PlutoRunner.jl:1000
#format_output#61@PlutoRunner.jl:1017[inlined]
#53@PlutoRunner.jl:909[inlined]
var"#with_io_to_logs#115"(::Bool, ::Base.CoreLogging.LogLevel, ::typeof(Main.PlutoRunner.with_io_to_logs), ::Main.PlutoRunner.var"#53#56"{Module, ShortCodes.DOI{String}, Dict{Tuple{UInt64, Int64}, Int64}})@PlutoRunner.jl:2424
(::Main.PlutoRunner.var"#119#120"{Bool, Base.CoreLogging.LogLevel, Main.PlutoRunner.var"#53#56"{Module, ShortCodes.DOI{String}, Dict{Tuple{UInt64, Int64}, Int64}}})()@PlutoRunner.jl:2448
with_logstate(::Function, ::Any)@logging.jl:511
with_logger@logging.jl:623[inlined]
#with_logger_and_io_to_logs#118@PlutoRunner.jl:2447[inlined]
var"#formatted_result_of#50"(::Bool, ::typeof(Main.PlutoRunner.formatted_result_of), ::Base.UUID, ::Base.UUID, ::Bool, ::Vector{String}, ::Nothing, ::Module)@PlutoRunner.jl:908
top-level scope@WorkspaceManager.jl:527
```

### PR DESCRIPTION
**Pull request by stefanbringuier**
## Summary

This PR adds a new function called `metadata_template(doi)` which returns a dictionary with fields corresponding to the OpenCitations metadata structure from the new API (https://w3id.org/oc/meta/api/v1/metadata). The function populates the values with null strings except for the `id` field, which is given the DOI`::String`

Additionally, the `fetch_metadata` function has been updated with an if-else clause that checks if the returned JSON from OpenCitations is empty for a given DOI. If empty, the new `metadata_template` function is called.

No changes were made to the `Base.show` methods, but the `year` function now uses a ternary operator condition to return null string if empty. This supports `Base.getproperty` call to `year`.

This PR addresses issue #13.

## Changes

- Added `metadata_template(doi)` function
- Updated `fetch_metadata` function with if-else clause for empty JSON
- Updated `year` function with ternary operator condition for `Base.getproperty`

## Testing

No new tests were added, but existing tests pass.

## Notes

Please review the changes and let me know if you have any modifications or design pattern changes.

### CODE REVIEW

**Comment by AshtonSBradley** [comment]
I hope I am testing this correctly. Running some of the examples from #13: 

Screenshot 2023-04-09 at 5 10 09 PM

Leads to, e.g. 

Screenshot 2023-04-09 at 5 11 04 PM

**Comment by stefanbringuier** [comment]
@AshtonSBradley Thanks for testing a few cases. I had set the `metadata_template` returned dictionary values to `null` character rather than an empty string, but the condition in `year` only checked for an empty string because in testing other DOIs there were some which showed metadata but with `pub_date` as an empty string.

I therefore updated the condition in the `year` function to check this. I tested this with Pluto and seems to work now: 
 

image

Let me know if you find anything else.

**Comment by AshtonSBradley** [comment]
This seems to solve it - errors are now links with cites.

**Comment by stefanbringuier** [comment]
> This seems to solve it - errors are now links with cites.

Do you have any proposed approach to format the display for the fallback when no data is present? I decided not to add a new `Base.show` method and to just fallaback on the existing methods.

**Comment by AshtonSBradley** [comment]
> > This seems to solve it - errors are now links with cites.
> 
> Do you have any proposed approach to format the display for the fallback when no data is present? I decided not to add a new `Base.show` method and to just fallaback on the existing methods.

It appears that the only data we have in these cases is the doi and the citation count. In some cases the doi is informative, e.g. in physical review journals. I would propose formatting the doi string as the url, followed by the citation count string, e.g:

`DOI("10.1103/PhysRevA.107.010101")`

currently formatted to

, () [10/cs5p5f](https://doi.org/10/cs5p5f), cited by 1931

would become

[10.1103/PhysRevA.107.010101](https://doi.org/10.1103/PhysRevA.107.010101), cited by 1931

**Comment by AshtonSBradley** [comment]
A related comment (off topic slightly, apologies), I wonder about simplifying the current formatting that provides a short link. Why not just mark up either the journal title or the article title to be the link, as done by some bibtex styles. What I mean is, e.g. currently:

`DOI("10.1098/rsta.2010.0333")` returns

Garraway, B M The Dicke Model In Quantum Optics: Dicke Model Revisited, Philosophical Transactions Of The Royal Society A: Mathematical, Physical And Engineering Sciences (2011) [10/c993k5](https://doi.org/10/c993k5), cited by 254

but this can be clarified and shortened to

Garraway, B M [The Dicke Model In Quantum Optics: Dicke Model Revisited](https://doi.org/10/c993k5), Philosophical Transactions Of The Royal Society A: Mathematical, Physical And Engineering Sciences (2011), cited by 254

or 

Garraway, B M The Dicke Model In Quantum Optics: Dicke Model Revisited,  [Philosophical Transactions Of The Royal Society A: Mathematical, Physical And Engineering Sciences (2011)](https://doi.org/10/c993k5), cited by 254

Personally I prefer the former as it makes the article content more evident

**Comment by stefanbringuier** [comment]
> ...
> would become
> 
> [10.1103/PhysRevA.107.010101](https://doi.org/10.1103/PhysRevA.107.010101), cited by 1931

I would agree this would look the best in my opinion. I'll wait to see if @hellemo wants to keep with this PR or go a different direction before making changes to the `Base.show` methods.

**Comment by hellemo** [comment]
Thanks for the PR, and sorry for not following up earlier.

I think this looks like a reasonable approach. One suggestion for a change I have is to use either an empty string `""` or perhaps `nothing` rather than the string `"\0"`.

For the suggested formatting change, I think that would be an improvement, so would be happy to accept a PR implementing that as well. For a bit of context, the idea of using the short doi was originally to have a very short textual (without hyperlink) reference to the paper, when copying from the pluto notebook to use in another document. But I guess very few potential readers would actually understand what to do with this short code, so linking the title seems more useful to me.

**Comment by stefanbringuier** [comment]
> I think this looks like a reasonable approach. One suggestion for a change I have is to use either an empty string `""` or perhaps `nothing` rather than the string `"\0"`.

No problems will implement.

> For the suggested formatting change, I think that would be an improvement, so would be happy to accept a PR implementing that as well. For a bit of context, the idea of using the short doi was originally to have a very short textual (without hyperlink) reference to the paper, when copying from the pluto notebook to use in another document. But I guess very few potential readers would actually understand what to do with this short code, so linking the title seems more useful to me.

For now, I'll just focus on handling the case where no metadata is present and return the full DOI with citations.

**Comment by AshtonSBradley** [comment]
Hi, just wondering if there is a problem with the missing metadata pr

File: src/doi.jl [hunk 1]
```diff
@@ -59,13 +59,27 @@ function Base.show(io::IO, ::MIME"text/html", dois::Array{T} where {T<:AbstractD
     print(io, "</ol>")
 end
 
+"""
+    meta-data structure from API: https://w3id.org/oc/meta/api/v1/metadata/
+"""
+function metadata_template(doi::String)
+    fields = (:publisher, :pub_date, :page, :venue, :issue, :editor, :author, :id, :volume, :title, :type)
+    rj = Dict(f => "\u0000" for f in fields)
```
Comment (hellemo) [inline src/doi.jl:67]:
```suggestion
    rj = Dict(f => "" for f in fields)
```
This was the change I had in mind, this should work as well, or is there a specific reason to use `"\u0000"`?

**Review by hellemo — CHANGES_REQUESTED** [review]
Looks good, is it ok to change to emty string, or am I missing something?

File: src/doi.jl (hunk 1 above)
Comment (stefanbringuier) [reply src/doi.jl:67]: This may work, but from my recollection, the `Base.show` methods didn't like all the fields being empty strings.

**Comment by AshtonSBradley** [comment]
Is there any reason this can't be merged? ShortCodes is unusable for doi fetching if it errors for incomplete metadata because any notebook with an incomplete return will appear to hang for about 2 minutes before returning an error

**Comment by hellemo** [comment]
> Is there any reason this can't be merged? ShortCodes is unusable for doi fetching if it errors for incomplete metadata because any notebook with an incomplete return will appear to hang for about 2 minutes before returning an error

Thanks for the reminder, it was the empty string thing, and I since forgot about it. Will merge and test and prepare a new release soon.

**Pull request merged**